app/master/cluster_master.py

```diff
@@ -143,6 +144,15 @@ def handle_slave_state_update(self, slave, new_slave_state):
         do_transition = slave_transition_functions.get(new_slave_state)
         do_transition(slave)
 
+    def _shutdown_slave(self, slave):
```

Comment:
I think we should be careful about making a clear distinction between "kill" and this new shutdown functionality. Like maybe we want to call it "graceful shutdown" or something like that?